cmd/kubectl-gadget/opensnoop.go

```diff
@@ -0,0 +1,156 @@
+
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/opensnoop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+	"github.com/spf13/cobra"
+)
+
+var opensnoopCmd = &cobra.Command{
+	Use:   "opensnoop",
+	Short: "Trace open() system calls",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		var transform func(string) string
+		switch params.OutputMode {
+		case utils.OutputModeJson: // don't print any header
+		case utils.OutputModeCustomColumns:
+			fmt.Println(getCustomopensnoopColsHeader(params.CustomColumns))
+			transform = formatEventopensnoopCostumCols
+		case utils.OutputModeColumns:
+			transform = opensnoopTransformLine
+			fmt.Printf("%-16s %-16s %-16s %-16s %-6s %-16s %-3s %3s %s\n",
+				"NODE", "NAMESPACE", "POD", "CONTAINER",
+				"PID", "COMM", "FD", "ERR", "PATH")
+		}
+
+		config := &utils.TraceConfig{
+			GadgetName:       "opensnoop",
+			Operation:        "start",
+			TraceOutputMode:  "Stream",
+			TraceOutputState: "Started",
+			CommonFlags:      &params,
+		}
+
+		err := utils.RunTraceAndPrintStream(config, transform)
+		if err != nil {
+			return fmt.Errorf("failed to run tracer: %w", err)
+		}
+
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(opensnoopCmd)
+	utils.AddCommonFlags(opensnoopCmd, &params)
+}
+
+func opensnoopTransformLine(line string) string {
+	var sb strings.Builder
+	var e types.Event
+
+	if err := json.Unmarshal([]byte(line), &e); err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling json: %s", err)
+		return ""
+	}
+
+	switch e.Type {
+	case eventtypes.NORMAL:
+		sb.WriteString(fmt.Sprintf("%-16s %-16s %-16s %-16s %-6d %-16s %-3d %3d %s",
+			e.Node, e.Namespace, e.Pod, e.Container,
+			e.Pid, e.Comm, e.Fd, e.Err, e.Path))
+
+		return sb.String()
+	case eventtypes.ERR:
+		fmt.Fprintf(os.Stderr, "error from node %s: %s", e.Node, e.Message)
+		return ""
+	}
+
+	return ""
+}
+
+func getCustomopensnoopColsHeader(cols []string) string {
+	var sb strings.Builder
+
+	for _, col := range cols {
+		switch col {
+		case "node":
+			sb.WriteString(fmt.Sprintf("%-16s", "NODE"))
+		case "namespace":
+			sb.WriteString(fmt.Sprintf("%-16s", "NAMESPACE"))
+		case "pod":
+			sb.WriteString(fmt.Sprintf("%-16s", "POD"))
+		case "container":
+			sb.WriteString(fmt.Sprintf("%-16s", "CONTAINER"))
+		case "pid":
+			sb.WriteString(fmt.Sprintf("%-6s", "PID"))
+		case "comm":
+			sb.WriteString(fmt.Sprintf("%-16s", "COMM"))
+		case "fd":
+			sb.WriteString(fmt.Sprintf("%-3s", "FD"))
+		case "err":
+			sb.WriteString(fmt.Sprintf("%-3s", "ERR"))
+		case "path":
+			sb.WriteString(fmt.Sprintf("%-24s", "PATH"))
+		}
+		sb.WriteRune(' ')
+	}
+
+	return sb.String()
+}
+
+func formatEventopensnoopCostumCols(line string) string {
+	var sb strings.Builder
+
+	var event types.Event
+	if err := json.Unmarshal([]byte(line), &event); err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling json: %s", err)
+		return ""
+	}
```

Comment:
This function is not managing the `eventtypes.ERR`. I think it is also missing for execsnoop and oomkill.

Reply:
It's missing all of them. opensnoopTransformLine() is missing DEBUG, INFO and WARN too. I think it's not scalable, I refactored the code to be a bit better but I still think there is a lot of duplication.